### PR TITLE
schemadiff: reject non-deterministic function in new column's default value

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

~~This PR introduces a new diff hint: `NonDeterministicDefaultStrategy`, which can take these values:~~

~~When it's set to `NonDeterministicDefaultReject`,~~

`schemadiff` now errors on table diffs where:

- A new column is added
- The column has an expression (non-literal) default value
- The expression contains a non-detereministic function, specifically one of:
  - `SYSDATE()`
  - `UUID()`
  - `RAND()`

This matches MySQL behavior as in:
```
alter table t add column v varchar(36) NOT NULL DEFAULT (uuid());
ERROR 1674 (HY000): Statement is unsafe because it uses a system function that may return a different value on the slave.
```

There is no need to handle `MODIFY COLUMN` or `CHANGE COLUMN` statements, because those cannot introduce un-computed column data, and only operate on existing columns. 

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/10203

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schemadiff/errors.go [hunk 1]
```diff
@@ -487,3 +487,13 @@ type PartitionSpecNonExclusiveError struct {
...
 }
+
+type NonDeterministicDefaultError struct {
+	Table    string
+	Column   string
+	Function string
+}
+
+func (e *NonDeterministicDefaultError) Error() string {
+	return fmt.Sprintf("column %s.%s default value uses non-deterministic function: %v", sqlescape.EscapeID(e.Table), sqlescape.EscapeID(e.Column), e.Function)
```
Comment (mattlord) [inline go/vt/schemadiff/errors.go:498]: Super nitty, but e.Function is a string so we can use %s instead of %v.

**Review by mattlord — APPROVED** [review]

File: go/vt/schemadiff/table.go [hunk 2]
```diff
@@ -1731,6 +1731,40 @@ func evaluateColumnReordering(t1SharedColumns, t2SharedColumns []*sqlparser.Colu
...
+				foundFunction = node.Name.String()
+				return false, nil
+			}
+		case *sqlparser.FuncExpr:
+			switch node.Name.Lowered() {
+			case "uuid", "rand":
+				foundFunction = node.Name.String()
+				return false, nil
+			default:
+				// recurse into function argument expressions
```
Comment (dbussink) [inline go/vt/schemadiff/table.go:1755]: Is this recursion needed? Walk already visits everything doesn't it?

File: go/vt/schemadiff/errors.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/errors.go]: Fixed.

File: go/vt/schemadiff/table.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/table.go]: Confirmed, the recursion is not needed. An earlier iteration led me to believe it was. The existing unit tests already cover two inner function scenarios so this is testes.

File: go/vt/schemadiff/table_test.go [hunk 3]
```diff
@@ -449,6 +452,88 @@ func TestCreateTableDiff(t *testing.T) {
 				"+	`y` int,",
 			},
 		},
+		{
+			name:  "added column with non deterministic expression, allow",
+			from:  "create table t1 (id int primary key, a int)",
+			to:    "create table t2 (id int primary key, a int, v varchar(36) not null default (uuid()))",
+			diff:  "alter table t1 add column v varchar(36) not null default (uuid())",
+			cdiff: "ALTER TABLE `t1` ADD COLUMN `v` varchar(36) NOT NULL DEFAULT (uuid())",
```
Comment (dbussink) [inline go/vt/schemadiff/table_test.go:460]: If this always errors on MySQL, should it even be a flag on schemadiff or should we also always use error?

Comment (shlomi-noach) [reply go/vt/schemadiff/table_test.go]: Agreed. Removed the option altogether, the issue title is now misleading and I'll fix it. Confirmed that MySQL rejects this kind of statement even for empty tables (as opposed to [another issue we found](https://github.com/vitessio/vitess/pull/16678) the other day where the success depends on whether the table is empty or not). This is a bit of inconsistent behavior for MySQL, but at least we're consistent with it...

**Review by dbussink — APPROVED** [review]

**Pull request merged**